Use incremental timeouts in spin_until_future_calls during BT execution to handle large server timeouts

naiveHobo opened the pull request:
Signed-off-by: Sarthak Mittal <sarthakmittal2608@gmail.com>

<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #2280 |
| Primary OS tested on | Ubuntu 20.04 |
| Robotic platform tested on | TurtleBot3 Gazebo simulation |

---

## Description of contribution in a few bullet points

* `BtServiceNode` and `BtActionNode` use incremental timeout values in all `rclcpp::spin_until_future_complete` calls while waiting for the server to response to a request when the specified server timeout exceeds the BT execution loop rate. This ensures that `rclcpp::spin_until_future_complete` don't slow down BT execution if the server timeout has to be increased for whatever reason.
* Expose `bt_loop_timeout` and `default_server_timeout` as node parameters.

## Description of documentation updates required from your changes

* Add `bt_loop_timeout` and `default_server_timeout` to navigation.ros.org

---

## Future work that may be required in bullet points


#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [x] Check that any significant change is added to the migration guide
- [x] Check that any new functions have Doxygen added
- [x] Check that any new features have test coverage
- [x] Check that any new plugins is added to the plugins page
- [x] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -257,7 +289,7 @@ class BtActionNode : public BT::ActionNodeBase
   /**
    * @brief Function to send new goal to action server
    */
-  void on_new_goal_received()
```
Please rename this function back to `on_new_goal_received`, to be in the styling of the rest of the file

naiveHobo replied:
Definitely. Though, just to confirm; this is a protected member function so it's not part of the API and probably doesn't need to follow the `on_` method name styling. I changed the name since now this function is only responsible for sending a new goal, whereas previously it was responsible for handling the complete new goal behavior. Should I revert it back to `on_new_goal_received`?

SteveMacenski replied:
That is a good point, I thought that was something overridable. I'd personally prefer to keep it the same name but I think your suggestion is also valid

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -174,7 +179,24 @@ class BtActionNode : public BT::ActionNodeBase
+    // if new goal was sent and action server has not yet responded
+    // check the future goal handle
+    if (goal_sent_) {
+      if (!check_future_goal_handle()) {
```
This logic will bypass on_wait_for_result which in turn will stop checks for path preemption in the FollowPath BT node. Is this block necessary if just below you have the same thing in L213-223?

naiveHobo replied:
The condition on L208 is true when there's a new goal that has to be sent to the action server while a previous goal is being executed. So L213-223 is used only to immediately check the future goal handle after a new goal has been sent **while a previous goal was being executed**. If `spin_until_future_complete` times out at this iteration, in the next iteration `goal_updated_` would be false and `check_future_goal_handle()` would never be called again. The L187-198 block is the main block that checks the future goal handle at every iteration if the future has not completed yet.

It is my understanding that `on_wait_for_result` is only meant to run when the BT node is waiting for the accepted action to complete, not when the BT node is waiting for the action server to respond to a `send_goal`. Is that right? In that case, `on_wait_for_result` should be bypassed until the future goal handle completes.

SteveMacenski replied:
> It is my understanding that on_wait_for_result is only meant to run when the BT node is waiting for the accepted action to complete, not when the BT node is waiting for the action server to respond to a send_goal. 

Thinking about this from a client-server perspective, I don't think the client (the BT node) cares about what the state of its request is in (sending goal, processing goal, etc) since for all it cares, once it sends, its being crunched. So I'd assume that would be called too any time its processing. Though knowing what FollowPath does, I'm not sure if having a new goal matters when its still sending the current goal - I don't think we can preempt that easily so maybe you're right we should keep it separate. 

I'm concerned about the `goal_sent_` variable - its not reset on failure / halt / terminal conditions. I think it could cause state to bleed into multiple requests. For instance, its only set back to false once accepted, but if rejected, it remains true which would impact the next request

naiveHobo replied:
Yup, that's right. I've missed a few important places where `goal_sent_` should have been reset.

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -49,10 +49,15 @@ class BtActionNode : public BT::ActionNodeBase
     getInput<std::chrono::milliseconds>("server_timeout", server_timeout_);
 
+    // Set spin_until_future_complete timeout
+    timeout_ = server_timeout_ < bt_loop_timeout_ ? server_timeout_ : bt_loop_timeout_;
```
`timeout_` is never really used. `server_timeout_` is in most places. I'm kind of wondering why this is around. If you just used `bt_loop_timeout_` in place of `timeout_` in L326, I don't think anything changes and removes this variable

naiveHobo replied:
> If you just used `bt_loop_timeout_` in place of `timeout_`

This could result in the total time that the BT node waits for exceeding the specified `server_timeout` value. Say if `bt_loop_timeout` is 10ms and `server_timeout` is 15ms, the node would wait for the server to timeout for 20ms.

SteveMacenski replied:
But you still do the subtraction to find if its larger, you just don't need this intermediary variable. `timeout_` is only used in this line https://github.com/ros-planning/navigation2/blob/5084f5617ae65c4973ced077f536263124d78d4e/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L326 which could be replaced as 

```
auto timeout = remaining > bt_loop_timeout_ ? bt_loop_timeout_ : remaining;
```

naiveHobo replied:
Ah yes, you're right!

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp:
```diff
@@ -91,6 +97,18 @@ bool BtActionServer<ActionT>::on_configure()
+  int timeout;
+  node->get_parameter("bt_loop_timeout", timeout);
+  bt_loop_timeout_ = std::chrono::milliseconds(timeout);
+  node->get_parameter("default_server_timeout", timeout);
```
Should this be `server_timeout` and `bt_loop_rate` (or `bt_loop_duration`)? The bt loop is used to set the running rate, so I'd think that's the best way to semantically describe it. Because we run at a certain rate, it relates to the timeout, but the timeout isn't it's main purpose (to regulate the BT tick rate). `server_timeout` is what we use in other places in the code base to describe the same thing as well

naiveHobo replied:
`bt_loop_duration` sounds good. `bt_loop_rate` suggests this parameter takes a frequency value.
`server_timeout` for an individual BT node can be overridden through the BT node input ports. If the user does not set the `server_timeout` input port on a BT node, it uses the default server timeout value that we put on the blackboard. `default_server_timeout` would make semantic sense to describe the actual purpose of this parameter. What do you think?

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -272,18 +304,44 @@ class BtActionNode : public BT::ActionNodeBase
+   * @brief Function to check response from the action server when a new goal is sent
+   * @return boolen True if future_goal_handle_ returns SUCCESS, False when it returns TIMEOUT
+   */
+  bool check_future_goal_handle()
```
Rename maybe to `if_goal_received`, I think semantically it helps explain what's happening here more and what a true vs a false return would actually mean. Maybe not exactly that name, but something similar?

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp:
```diff
@@ -107,9 +109,13 @@ class BtServiceNode : public BT::SyncActionNode
    */
   BT::NodeStatus tick() override
   {
-    on_tick();
```
I'm debating this question - do we actually _need_ to change the service node at all? These should be short running computations and the `future_result_` doesn't give you an accepted / declined feedback the way actions do to wait until accepted before continuing (since that ack would actually be the response itself).

naiveHobo replied:
I agree. The only thing is that it cloud potentially slow down the tick rate if say a server takes some time to respond or the service fails for whatever reason. That being said, I don't think these changes to the BT service nodes would affect the performance in any way, might as well add them to avoid any potential issues in the future?

SteveMacenski replied:
OK, sounds good. I looked over the implementation and looks good to me

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -313,6 +371,15 @@ class BtActionNode : public BT::ActionNodeBase
+  // To track the action server response when a new goal is sent
+  std::shared_future<typename rclcpp_action::ClientGoalHandle<ActionT>::SharedPtr>
+  future_goal_handle_;
+  bool goal_sent_{false};
```
I think some of my confusion in reading this is just naming, some of the variables aren't as descriptive of what they're being used for as I'd like (`goal_sent_` `check_future_goal_handle` come to mind) or what the return of true/false would mean by the function name.

SteveMacenski commented:
Should as part of this PR we increase the action send_goal timeout so that we remove the issue people are seeing? How high should we make it?

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -174,7 +176,25 @@ class BtActionNode : public BT::ActionNodeBase
+
+    // if new goal was sent and action server has not yet responded
+    // check the future goal handle
+    if (goal_sent_) {
```
Could `goal_sent_` be replaced by a check on `goal_handle_` instead? If its `reset()` after each goal, then it should be a nullpointer while the goal request is being processed without needing to explicitly track this state variable. 

We'd need to make sure in `halt` and all other terminal conditions that the `goal_handle_` is `reset()` so that its only valid when no goal is sent yet or its being sent currently.

naiveHobo commented:
> Should as part of this PR we increase the action send_goal timeout so that we remove the issue people are seeing? How high should we make it?

I've not been able to reproduce the timeout issues that people are seeing. I'm thinking the timeout would be dependent on the system specifications so difficult to come up with a single value that works every time. 100ms server timeout should be large enough to account for most systems right?

SteveMacenski commented:
> 100ms server timeout should be large enough to account for most systems right?

Yeah, that seems super high. I would say 20-50ms. One measurement a user made said they were seeing as high as 18ms on their highly loaded system. So 20 I think would cover it and also only slow the BT by at most 1 tick response. Though I like "final" solutions, so going up to 40ms or 50ms seems like a way to just put this to bed (100ms seems like its high enough that might be a real failure to me if the network can't process a request it should be able to do in 10 nanoseconds in 100 milliseconds).

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -272,18 +316,38 @@ class BtActionNode : public BT::ActionNodeBase
       };
 
-    auto future_goal_handle = action_client_->async_send_goal(goal_, send_goal_options);
+    goal_handle_.reset();
```
I don't think this reset is required, its about to be overridden which implicitly destructs the old object

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -272,18 +316,38 @@ class BtActionNode : public BT::ActionNodeBase
+    goal_handle_.reset();
+    future_goal_handle_ = action_client_->async_send_goal(goal_, send_goal_options);
+    time_goal_sent_ = node_->now();
+    time_elapsed_since_goal_sent_ = std::chrono::milliseconds(0);
```
I'd argue sending the `elapsed_time` as a function input is easier to logically track, but this also works

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -272,18 +316,38 @@ class BtActionNode : public BT::ActionNodeBase
+    auto timeout = server_timeout_ > bt_loop_duration_ ? bt_loop_duration_ : server_timeout_;
+    auto result = rclcpp::spin_until_future_complete(node_, future_goal_handle_, timeout);
+
+    time_elapsed_since_goal_sent_ =
```
Shouldn't this be computed very first in the `tick`? Also I'm looking at `timeout = ...` and there's no looking at elapsed if less than `server_timeout_` or `bt_loop_duration_` duration remains to spin only the remainder.  I believe you had that before using `remaining` time.

SteveMacenski commented:
If there's some test we can add to show that this works the way we expect it to with timeouts < loop duration and timeouts > loop duration with a server that blocks for < loop duration and > loop duration, that would be great to make sure. I think we cover the bases here, but this is a subtle thing to reset the state on and I want to be sure that this works as expected, and with multiple requests, if at all possible

KavenYau commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -272,18 +318,44 @@ class BtActionNode : public BT::ActionNodeBase
         }
```
Should't we check if the `goal_handle_` is `nullptr`? because it seems that you call `goal_handle_.reset()` in the timeout handling segment.

SteveMacenski replied:
If we get a `result`, there shouldn't be any situation where the handle is a nullpointer, this is a callback on success / failure / cancel. If its a nullpointer, we have an issue where we can't then get the `get_goal_id()` in order to validate that this is the proper response to set the flags that the action has completed to return a result to the behavior tree in `tick()`. 

If this is an issue, we need to rethink how we deal with the goal handles in the bt action node.

The `goal_handle_` I don't think is set to anything when it is reset due to timeouts. The future to the goal handle (https://github.com/ros-planning/navigation2/blob/f83612d28ce0db76436cf711071129eb17099c8f/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L321) is still spinning to get valid. The `goal_handle_` from the future isnt set until it spins and gets back a successful goal acceptance (https://github.com/ros-planning/navigation2/blob/f83612d28ce0db76436cf711071129eb17099c8f/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L351) so I think in (all?) most cases, that should be a nullpointer anyway.

@naiveHobo if that is correct, should the checks on `goal_handle_` in some places actually be `future_goal_handle_` and we reset `future_goal_handle_` on timeout? I think we may have been mixing these up at various points. Should it actually be:

```
    if (future_goal_handle_) {
      auto elapsed = (node_->now() - time_goal_sent_).to_chrono<std::chrono::milliseconds>();
      if (!is_future_goal_handle_complete(elapsed)) {
        // return RUNNING if there is still some time before timeout happens
        if (elapsed < server_timeout_) {
          return BT::NodeStatus::RUNNING;
        }
        // if server has taken more time to respond than the specified timeout value return FAILURE
        RCLCPP_WARN(
          node_->get_logger(),
          "Timed out while waiting for action server to acknowledge goal request for %s",
          action_name_.c_str());
        future_goal_handle_.reset();
        return BT::NodeStatus::FAILURE;
      }
    }
```

?? We track and reset the `future_goal_handle_` instead? Before we set, we should default the `future_goal_handle_` to nullpointer. Then if we send a goal, it becomes non-null to then spin the future. Once the future is complete (or times out or fails) we should reset the `future_goal_handle_` instead back to nullpointer? I'm not sure anything relating to `send_goal` and `is_future_goal_handle_complete` needs `goal_handle_`, I think it needs `future_goal_handle_` checks, yeah?

SteveMacenski replied:
I think if we remove the `goal_handle_.reset()`'s for this logic in `tick()` and `is_future_goal_handle_complete()`, then the only remaining ones are in `halt` and the one at the end of `tick()` once a final solution is available. **Would that make things work well?** The other 4 could be the `future_goal_handle_.reset()` I think...

(this is overall a subtle issue so correct me if I missed something)

SteveMacenski replied:
At the end of `tick()` I agree, because we wouldn't have gotten there without a `result` in that callback. Though at the end of `halt()` are we _guaranteed_ to have received a result? I'm actually not 100% sure about that. Does the `spin_until_future_complete` spin until the action server accepts the cancellation or until the cancellation is actually complete? 

If it just waits for the cancel to be accepted, in the simple action server (https://github.com/ros-planning/navigation2/blob/9f0e85f312770a63eeac5b37fc4b8745c38187d3/nav2_util/include/nav2_util/simple_action_server.hpp#L135-L149) that doesn't promise that we have actually completed that cancel (I think that happens in https://github.com/ros-planning/navigation2/blob/9f0e85f312770a63eeac5b37fc4b8745c38187d3/nav2_util/include/nav2_util/simple_action_server.hpp#L523 `canceled`, probably)

As such, maybe its best for the goal handle reset to be in end of tick and in the result callback after the quantities are set if the goal IDs match?

naiveHobo replied:
Yup, got it. I'll remove the reset at the end of `halt()`. Although, I don't think we should reset the goal handle in the result callback. It might be useful to have access to the last goal through the goal handle in the `on_success()`, `on_aborted()`, `on_cancelled()` functions after the goal result is available. The reset at the end of `tick()` should ensure that the goal handle is always reset after a goal.

SteveMacenski commented:
```
/opt/overlay_ws/src/navigation2/nav2_behavior_tree/test/plugins/action/test_bt_action_node.cpp:328
Expected equality of these values:
  ticks
    Which is: 6
  9
```

One of the new tests are failing - actually both of the testing runs had different values (6 vs 8)

---

I'm happy with this overall, I don't think I see any other issues needing to be addressed. In light of that, @tzskp1 and @KavenYau might want to take a look / test again to see if it resolves the issues in #2305. However, since those are currently existing issues, as long as this doesn't make it any **worse**, I'm inclined to merge this since we still don't know root cause of those issues and this is supposed to fix something non-trivial. 

@naiveHobo however, his PR should make the CI not flaky anymore, and I'm seeing alot of failures including

```
[bt_navigator-10] [WARN] [1620137230.106779840] [bt_navigator_rclcpp_node]: Timed out while waiting for action server to acknowledge goal request for compute_path_to_pose
[planner_server-8] [INFO] [1620137230.119440375] [global_costmap.global_costmap]: Received request to clear entirely the global_costmap
[bt_navigator-10] [ERROR] [1620137230.122825667] [BehaviorTreeEngine]: Behavior tree threw exception: SyncActionNode MUST never return RUNNING. Exiting with failure.
[bt_navigator-10] [ERROR] [1620137230.131459446] [bt_navigator]: Goal failed
[bt_navigator-10] [WARN] [1620137230.132364221] [bt_navigator]: [navigate_to_pose] [ActionServer] Aborting handle.
[tester_node-13] [INFO] [1620137230.134366128] [nav2_tester]: ?[1;37;44mGoal failed with status code: 6?[0m
[tester_node-13] [ERROR] [1620137230.134978606] [nav2_tester]: ?[1;37;41mTest FAILED?[0m
```

Which is basically the new version of the `send_goal failed` error (and with a new really not good error of that exception from the BT Engine). I think there's something still wrong here -- certainly with state tracking for that new exception -- but also that it didn't fix the error caused by the goal request timing out (did you update the timeout from 10ms to 20ms?)

naiveHobo commented:
Not quite sure what the issue is now, I've updated all the server timeouts to 20ms. In the tests I added for the BT action node changes, the number of ticks before timeout varies (most likely due to ros clock timing) but I can confirm that the total time that the bt action node waits for the server to acknowledge is always equal to the server timeout value (~1-2ms).

SteveMacenski commented:
One of the tests is still failing and from the tests that are failing 
```

[tester_node-13] [INFO] [1620163729.442322811] [nav2_waypoint_tester]: Sending goal request...
[tester_node-13] [ERROR] [1620163729.444101914] [nav2_waypoint_tester]: Goal rejected
```

It seems like this has some issues where goals are being rejected now? I don't believe this is happening in main, but I'm not 100% sure of that either. 

The release test otherwise succeeded other than the bt action server new test that genuinely looks to have failed

naiveHobo commented:
> It seems like this has some issues where goals are being rejected now?

That nav2_waypoint_tester failure you mentioned seems unrelated to this PR, it happened due to a gazebo server failure I believe:

```
[gzserver-1] gzserver: /usr/include/boost/smart_ptr/enable_shared_from_this.hpp:51: boost::shared_ptr<X> boost::enable_shared_from_this<T>::shared_from_this() [with T = gazebo::transport::Node]: Assertion `p.get() == this' failed.
[ERROR] [gzserver-1]: process has died [pid 12244, exit code -6, cmd 'gzserver -s libgazebo_ros_init.so --minimal_comms /opt/overlay_ws/src/navigation2/nav2_system_tests/worlds/turtlebot3_ros2_demo.world'].
```

I just retriggered the CI and everything passed.

SteveMacenski commented:
One day... one day soon we'll have non-flaky CI again... I keep telling myself to sleep at night :wink:

SteveMacenski reviewed (approved):
I'm happy with this! Thanks for all the hard work, this will be really nice to have resolved going into Galactic release :-) 

I'm letting CI to run again but assuming only issues outside of the scope of this work are found, I'll merge. @tzskp1 please test this, it does make about 70% of the changes your PR requests (except for the result callback checks) that might actually resolve the issue you see.

If it does not, we can address your PR separately

SteveMacenski commented:
@naiveHobo I jumped the gun a little, can you submit a PR to update the docs for the new parameter(s) this added?
